File: leetcode/dp/grid/0221_maximal_square.py
# LC 221. Maximal Square | Medium
class Solution:
    def maximalSquare(self, matrix):
        m, n = len(matrix), len(matrix[0]); mx = 0
        dp = [0] * (n + 1)
        for i in range(m):
            new_dp = [0] * (n + 1)
            for j in range(1, n + 1):
                if matrix[i][j-1] == '1':
                    new_dp[j] = min(dp[j], dp[j-1], new_dp[j-1]) + 1
                    mx = max(mx, new_dp[j])
            dp = new_dp
        return mx * mx

File: leetcode/dp/grid/0221_maximal_square.cpp
// LC 221. Maximal Square | Medium
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size(), n = matrix[0].size(), mx = 0;
        vector<int> dp(n+1, 0);
        for (int i = 0; i < m; i++) {
            vector<int> ndp(n+1, 0);
            for (int j = 1; j <= n; j++)
                if (matrix[i][j-1] == '1') { ndp[j] = min({dp[j], dp[j-1], ndp[j-1]})+1; mx = max(mx, ndp[j]); }
            dp = ndp;
        }
        return mx*mx;
    }
};
